client/lib/url/url-type.ts

```diff
@@ -0,0 +1,78 @@
+/**
+ * External dependencies
+ */
+import { URL as URLString } from 'types';
+import { Falsey } from 'utility-types';
+
+export enum URL_TYPE {
+	// A complete URL, with (at least) protocol and host.
+	// E.g. `http://example.com` or `http://example.com/path`
+	ABSOLUTE = 'ABSOLUTE',
+	// A URL with no protocol, but with a host.
+	// E.g. `//example.com` or `//example.com/path`
+	PROTOCOL_RELATIVE = 'PROTOCOL_RELATIVE',
+	// A URL with no protocol or host, but with a path starting at the root.
+	// E.g. `/` or `/path`
+	ROOT_RELATIVE = 'ROOT_RELATIVE',
+	// A URL with no protocol or host, but with a path relative to the current resource.
+	// E.g. `../foo` or `bar`
+	PATH_RELATIVE = 'PATH_RELATIVE',
+	// Any invalid URL.
+	// E.g. `///`
+	INVALID = 'INVALID',
+}
+
+const BASE_HOSTNAME = '__domain__.invalid';
+const BASE_URL = `http://${ BASE_HOSTNAME }`;
+
+/**
+ * Determine the type of a URL, with regards to its completeness.
+ * @param url the URL to analyze
+ *
+ * @returns the type of the URL
+ */
+export function determineUrlType( url: URLString | URL | Falsey ) {
```

Comment:
Shall we add an explicit return type? I'd expect this always to return a `URL_TYPE`:

```suggestion
export function determineUrlType( url: URLString | URL | Falsey ): URL_TYPE {
```

Reply:
I didn't add it explicitly because TypeScript was able to determine it implicitly, but I'm happy making it explicit 👍